framework/validators/IpValidator.php

```diff
@@ -0,0 +1,498 @@
+                if ($cidr > static::IPV4_ADDRESS_LENGTH || $cidr < 0) {
+                    return [$this->wrongCidr, []];
+                }
+            } else {
+                $isCidrDefault = true;
+                $cidr = static::IPV4_ADDRESS_LENGTH;
+            }
+
+            if (!$this->ipv4) {
+                return [$this->ipv4NotAllowed, []];
+            }
+            if (!$this->validateIPv4($ip)) {
+                return [$this->wrongIp, []];
+            }
+        }
+
+        if (!$this->isAllowed($ip, $cidr)) {
+            return [$this->notInRange, []];
+        }
+
+        $result = $negation . $ip;
+
+        if ($this->subnet !== false && (!$isCidrDefault || $isCidrDefault && $this->normalize)) {
+            $result .= "/$cidr";
+        }
+
+        return $result;
+    }
+
+    /**
+     * Expands an IPv6 address to it's full notation. For example `2001:db8::1` will be
+     * expanded to `2001:0db8:0000:0000:0000:0000:0000:0001`
+     *
+     * @param string $ip the original IPv6
+     * @return string the expanded IPv6
+     */
+    private function expandIPv6($ip)
+    {
+        $hex = unpack('H*hex', inet_pton($ip));
+        return substr(preg_replace('/([a-f0-9]{4})/i', '$1:', $hex['hex']), 0, -1);
+    }
+
+    /**
+     * The method checks whether the IP address is allowed according to the [[ips]] lists
+     *
+     * @param string $ip
+     * @param integer $cidr
+     * @return boolean
+     * @see ips
+     */
+    private function isAllowed($ip, $cidr)
+    {
+        $denied = false;
+        $allowed = true;
+        if (!empty($this->ips['deny']) && $this->inRange($ip, $cidr, $this->ips['deny'])) {
+            $denied = true;
+        }
+        if (!empty($this->ips['allow']) && !$this->inRange($ip, $cidr, $this->ips['allow'])) {
+            $allowed = false;
+        }
+
+        return current(array_keys($this->ips)) === 'allow' ? (!$denied && $allowed) : (!$denied || $allowed);
+    }
+
+    /**
+     * Validates IPv4 address
+     *
+     * @param string $value
+     * @return boolean
+     */
+    protected function validateIPv4($value)
+    {
+        return preg_match($this->ipv4Pattern, $value) !== 0;
+    }
+
+    /**
+     * Validates IPv6 address
+     *
+     * @param string $value
+     * @return boolean
+     */
+    protected function validateIPv6($value)
+    {
+        return preg_match($this->ipv6Pattern, $value) !== 0;
+    }
+
+    /**
+     * Gets the IP version
+     *
+     * @param string $ip
+     * @return integer
+     */
+    private function getIpVersion($ip)
+    {
+        return strpos($ip, ':') === false ? 4 : 6;
+    }
+
+    /**
+     * Used to get the Regexp pattern for initial IP address parsing
+     * @return string
+     */
+    private function getIpParsePattern()
+    {
+        $negationChar = is_string($this->negationChar) ? $this->negationChar : static::DEFAULT_NEGATION_CHAR;
+        return '/^(' . $negationChar . '?)(.+?)(\/(\d+))?$/';
```

Comment:
DEFAULT_NEGATION_CHAR goes into regexp not escaped.
But it still works properly because it is followed by '?'.
I think preg_quote should be moved here from setNegationChar() which makes get/setNegationChar unnecessary.